src/java.base/share/classes/jdk/crac/impl/AbstractContextImpl.java

```diff
@@ -55,32 +59,44 @@ protected synchronized void register(R resource, P payload) {
 
     @Override
     public synchronized void beforeCheckpoint(Context<? extends Resource> context) throws CheckpointException {
-        List<R> resources = checkpointQ.entrySet().stream()
-            .sorted(comparator)
-            .map(Map.Entry::getKey)
-            .collect(Collectors.toList());
-
-        CheckpointException exception = new CheckpointException();
-        for (Resource r : resources) {
-            if (DEBUG) {
-                System.err.println("jdk.crac beforeCheckpoint " + r.toString());
+        synchronized (enterCriticalSectionLock) {
+            synchronized (exitCriticalSectionLock) {
+                while (!lockQ.isEmpty()) {
```

Comment:
This use of `lockQ` is not synchronized. I assume there is no race with add-to/remove-from lockQ, since they are synced with enterCriticalSectionLock/exitCriticalSectionLock as well. But a comment will be useful.

Is it possible to remove enterCriticalSectionLock, exitCriticalSectionLock to make code more simple? And use a single lock to sync the list, to block entering ot checkpoint-critical section, and to notify the checkpointRestore the last section is finished. Or is it so to prioritize checkpoint over checkpoint-critical sections?

Reply:
Thank you. You are right. It could be simplified. It is not possible to use only one lock but I've merged enterCriticalSectionLock and exitCriticalSectionLock locks into one criticalSectionsLock